tt_metal/jit_build/kernel_args.cpp

```diff
@@ -0,0 +1,58 @@
+// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include <regex>
+#include <map>
+#include <mutex>
+#include <fstream>
+#include "common/utils.hpp"
+
+using namespace std;
+
+namespace tt::tt_metal {
+
+// Maps full_kernel_name to defines used to compile it
+map<string, string> kernel_defines_and_args_; // Maps full_kernel_name to defines used to compile it
+// Guards kernel_defines_and_args_ for multi-threaded access
+std::mutex mutex_kernel_defines_and_args_;
+
+// Replace all occurrences of 'from' in 'source' with 'to'
+void str_replace_all(string& source, const string& from, const string& to) {
+    for(size_t pos = 0; (pos = source.find(from, pos)) != string::npos; pos += to.length()) {
+        source.replace(pos, from.length(), to);
+    }
+}
+
+void log_kernel_defines_and_args (const string &out_kernel_root_path, const string &full_kernel_name, const string &built_in_defines, const string &defines_and_args_str) {
+    std::lock_guard<std::mutex> lock(mutex_kernel_defines_and_args_);
+    string defines_as_csv(built_in_defines + " " + defines_and_args_str);
+    str_replace_all(defines_as_csv, "KERNEL_COMPILE_TIME_", "");
+    str_replace_all(defines_as_csv, "-D", ",");
+    str_replace_all(defines_as_csv, " ", "");
+
+    if (kernel_defines_and_args_.find(full_kernel_name) == kernel_defines_and_args_.end()) {
+        kernel_defines_and_args_[full_kernel_name] = defines_as_csv;
+    } else {
+        // TT_ASSERT(kernel_defines_and_args_[full_kernel_name] == defines_as_csv, "Mismatched defines and args for kernel: {}. Expected:\n{}, Got:     \n{}", full_kernel_name, kernel_defines_and_args_[full_kernel_name], defines_as_csv);
+    }
+}
+
+void dump_kernel_defines_and_args(const string &out_kernel_root_path) {
+    // Make sure the directory exists
+    tt::utils::create_file(out_kernel_root_path);
+
+    string kernel_args_csv = out_kernel_root_path + "kernel_args.csv";
+    ofstream file(kernel_args_csv, ios::trunc);
+    if (file.is_open()) {
+        std::lock_guard<std::mutex> lock(mutex_kernel_defines_and_args_);
```

Comment:
what's this lock guarding against?  should it be above the file open above?  if 2 threads can hit here, seems the 2nd could still clobber the first by truncating the file?

Reply:
It is against the kernel_defines_and_args_ map (it is shared across threads). You are right about the file, I will move the mutex to guard the file as well.